Kata 2.0: add metrics design documents (#665)
kata-containers/documentation: how-to/how-to-set-prometheus-in-k8s.md

# How to monitor Kata Containers in K8s clusters

> This document will describe how to run `kata-magent` in a K8s cluster, and using Prometheus's K8s service discovery to scrape metrics from `kata-agent`.

Reviewer: ```suggestion
This document describes how to run `kata-magent` in a Kubernetes cluster using Prometheus's service discovery to scrape metrics from `kata-agent`.
```

- [Introduction](#introduction)
- [Pre-requisites](#pre-requisites)
- [Configure Prometheus](#configure-prometheus)
- [Configure `kata-magent`](#configure-kata-magent)

> **Warning**: This how-to is only for evaluation purpose, you **SHOULD NOT** running it in production using this configurations.

## Introduction

If you are running Kata containers in a K8s cluster, the best way to run `kata-magent` is using K8s native `DaemonSet`, `kata-magent` will run on desired K8s nodes without other operations when new nodes joined the cluster.

Prometheus also support a K8s service discovery that can find scrape targets dynamically without explicitly setting `kata-magent`'s metric endpoints.

## Pre-requisites

> You must have a running K8s cluster first. If not, [install a K8s cluster](https://kubernetes.io/docs/setup/) first.

Reviewer: Should we spell out K8s -> Kubernetes throughout the doc?

Should we highlight that a Kata compatible CRI needs to be utilized as well (containerd or crio)?

Author: About the second question, a **working K8s cluster** implicitly say that **working CRI(containerd or others)** is included, I think.

Insteadly, I can add links about how to run  Kata Containers with K8s/containerd.

Also you should ensure that `kubectl` working correctly.

## Configure Prometheus

> Running Prometheus is easy.

Reviewer: I agree its easy, but let's not qualify the ease of use, as for some people its trivial, and for other people it'll be the hardest thing they ever do. :D

```suggestion
Start Prometheus by utilizing our sample manifest:  
```

```
$ kubectl apply -f https://raw.githubusercontent.com/kata-containers/documentation/master/how-to/data/prometheus.yml
```

> This will create a new namespace named `prometheus`, and create some resources for Prometheus:

Reviewer: ```suggestion
This will create a new namespace, `prometheus`, and create the following resources:
```

* `ClusterRole`, `ServiceAccount`, `ClusterRoleBinding` to let Prometheus to access K8s API server.
* `ConfigMap` that contains minimum configurations to let Prometheus run K8s service discovery.
* `Deployment` that run Prometheus in `Pod`.
* `Service` with `type` of `NodePort`(`30909` in this how to), that we can access Prometheus through `<hostIP>:30909`. In production environment, this `type` may be `LoadBalancer` or `Ingress` resource.

> 

Reviewer: nit: extra line removal

> After the Prometheus server has running, open `hostIP:NodePort` in browser to verify Prometheus is working correctly.

Reviewer: Optional: should we describe how to determine it is ready/running?
```suggestion
After the Prometheus server is running, open `hostIP:NodePort` in browser to verify Prometheus is working correctly.
```

Author: Mostly we can open the URL, that is saying the Prometheus is working correctly.

I changed it to a more common routine by issue `curl` command.

## Configure `kata-magent`

> As Prometheus, running `kata-magent` need only one command too.

Reviewer: ```suggestion
`kata-magent` can be started on the cluster as follows:
```

```

> $ kubectl apply -f https://raw.githubusercontent.com/kata-containers/documentation/master/how-to/data/kata-magent-daementset.yml

Reviewer: We should update the name of the URL and file: s/daemen/daemon

```

> This will create a new namespace `kata-system` and a deployment in it.

Reviewer: It is actually a daemonset, right?

As an aside from this doc PR: do you think we should move kata-deploy into this namespace as well (and future operator)?

Author: Yes, it's a daemonset.

Vendor product should run in its own namespace, and Kata containers workload also should run in the same namespace.

Reviewer: OK, can you s/deployment/daemonset ?

Author: Done.

> After the pod is becoming running, Prometheus should have discovered `kata-magent` as a target. You can open `http://<hostIP>:30909/service-discovery` and find `kubernetes-pods` under the `Service Discovery` list, if there have no problems.

Reviewer: ```suggestion
Once the daemonset is running, Prometheus should discover `kata-magent` as a target. You can open `http://<hostIP>:30909/service-discovery` and find `kubernetes-pods` under the `Service Discovery` list
```

# How to monitor Kata Containers in Kubernetes clusters

> This document describes how to run `kata-magent` in a Kubernetes cluster using Prometheus's service discovery to scrape metrics from `kata-agent`.

Reviewer 2: I still don't like the name `kata-magent` I'm afraid :smile:

Anyone else have thoughts on this?

Reviewer 3: @jodh-intel How about `kata-monitor` or `kata-monitoring-agent`, those are the two that I could think of. Now that think about it more, I dont mind the the longer name `kata-monitoring-agent`.

@egernst @fidencio @bergwolf Please chip in with any suggestions.

Reviewer 2: I think I prefer `kata-monitor` - it not too long, but suggests what the component is doing. I'd rather we didn't "over-load" terms to avoid confusion (in other words, we already have one agent, so adding another is bound to confuse :-)

Reviewer: I’m okay with all of the suggestions above. Spelling out the M is better imo?

Reviewer 4: I'd prefer something like kata-monitor, and then document properly what's monitoring. :-)

Reviewer 5: +1 for `kata-monitor`

Author: It seems the final winner is `kata-monitor`, I have updated.
